Material Canvas: Basic matrix node configs and vector casting improvements (mostly JSON changes)

gadams3 opened the pull request:
## What does this PR do?

Created basic node configurations for floating point matrix types. Support for connecting to these data types will be added to other nodes after another change that adds more intuitive settings for specifying supported scalar, vector, matrix data types and sizes.

Extended all of the color and vector variable and input nodes with XYZW/RGBA output slots as a convenience to give immediate access to all of the vector and color components.

Updated experimental PBR template and a couple of notes to use explicit C style casting so that scalars and vectors can be converted to the correct type according to HLSL specifications. I was previously attempting to do this with implicit type conversions or accessing vector elements. The C style casting handles the cases uniformly and stops implicit type conversion warnings.

Resolves https://github.com/o3de/o3de/issues/12812

## How was this PR tested?

![image](https://user-images.githubusercontent.com/82461473/198641281-9717ad3b-507c-40eb-b351-8ca3190716ce.png)
![image](https://user-images.githubusercontent.com/82461473/198641965-544fe8f1-d069-4822-a7fa-5d5c233ef809.png)
![image](https://user-images.githubusercontent.com/82461473/198643023-c8731fdb-6e70-413c-b48b-7482e56dad45.png)

onecent1101 commented on Gems/Atom/Tools/MaterialCanvas/Assets/MaterialCanvas/TestData/Nodes/Functions/clamp.materialcanvasnode:
```diff
@@ -1,4 +1,4 @@
-{
```
extra back slash?

gadams3 replied:
Oh no. I missed that. Submitting a fix

gadams3 replied:
https://github.com/o3de/o3de/pull/12889

onecent1101 reviewed (approved):
Should we move the node configuration files into proper directory? they are under `TestData`, the name is so confusing

gadams3 commented:
> Should we move the node configuration files into proper directory? they are under `TestData`, the name is so confusing

Technically, this is still an experimental release, moving into preview. I make periodic changes to the settings and formats. Will be adding better descriptions, display names, and so on. At least for the next few weeks, I want to continue calling it test data. All of the node configurations used to be referenced by paths but I changed them to be registered and looked up by UUID so the files can be moved, renamed, reorganized at any point.
